apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/UnchainedTraceHeaderPropagatingHandler.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// Class for tracing outgoing HTTP requests and propagating the trace header.
+    /// </summary>
+    /// <remarks>
+    /// Ensures the trace header is propagated in the headers for outgoing HTTP requests and 
+    /// traces the total time of the outgoing HTTP request. This is only done if tracing is initialized
+    /// and tracing is enabled for the request current request.
+    /// 
+    /// Explicitly leaves unset the InnerHandler property so that this Handler can be used with the
+    /// System.Net.Http.IHttpClientFactory defined in Microsoft.Extensions.Http.
+    /// </remarks>
+    public class UnchainedTraceHeaderPropagatingHandler : DelegatingHandler
```

Comment:
I'm not sure about "unchained" in the name here. Let's mull over some better names.

Reply:
Yep, I don't like unchained very much either.

Reply:
Thinking about this further, do we actually need two classes? Could we just have two constructors, one with a handler and one without? Then just document the situation for each of them?

It feels a bit odd to have a subclass just for the sake of setting a property. Or is it needed for dependency injection?

Reply:
The problem is that the `MessageHandler` parameter in the existing constructor is optional, a lot of our own code uses that, expecting the `InnerHandler` to be set to `new MessageHandler()` if the parameter is not, and potentially some user code as well. So we cannot add a new constructor that doesn't take a MessegaHandler, unless we want to go through a breaking change that will be breaking for us as well. And if we go with that we'd be forcing client code to always pass a `MessageHandler` even if they are happy with the "default" `new MessageHandler()`, which by the way, is always necessary except on this new client factory use case.
We can't even keep the existing constructor and add a static creation method `WithoutInnerHandler` because the existing constructor always sets InnerHandler, which cannot be set back to null later (I tried all of that before deciding on the two classes).

Reply:
Hmmm, yes, a private constructor with a dummy will work.
Name wise, for the static creation method, I like better `WithoutInnerHandler` or `NoInnerHandler` because it describes the instance instead of the usage, which is now only for HttpClientFactory, but might be for other things as well in the future? But my heart is not set on that so let me know what you prefer.

I'll try the private constructor approach in a separate commit, let's see how that looks.

Reply:
> Presumably we could have two different calls and only register one... but that's probably overly brittle.

I don't get what you mean.

> am I right in saying that user code doesn't need to see it, because it just gets registered by our trace setup methods?

No, unfortunately users do get to see it, because they have to configure the HttpClient to use that particular message handler, we configure it for injection so they can use it when configuring the HttpClient.